File: 2011.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	const DIV = 1000000
	r := bufio.NewReader(os.Stdin)
	input, _ := r.ReadString('\n')
	enc := strings.TrimSpace(input)
	l := len(enc)
	dp := make([]int, l+1)
	if enc[l-1] != '0' {
		dp[l-1] = 1
	}
	dp[l] = 1
	for i := l - 2; i >= 0; i-- {
		beforeTwoNum, _ := strconv.Atoi(string(enc[i]) + string(enc[i+1]))
		if beforeTwoNum >= 10 && beforeTwoNum <= 26 {
			dp[i] = (dp[i+1] + dp[i+2]) % DIV
		} else if beforeTwoNum < 10 {
			continue
		} else {
			dp[i] = dp[i+1]
		}
	}
	fmt.Printf("%d\n", dp[0])
}

File: 2011.py
string = input()

def is_valid_number(i, j):
    if i == j:
        return string[i] != '0'

    if string[i] == '0' or string[i] > '2':
        return False

    if string[i] == '2' and string[j] > '6':
        return False

    return True

dp = [ 0 for _ in range(len(string)) ]

for i in range(len(string)):
    if i == 0 and is_valid_number(0,0):
        dp[i] = 1
    elif i == 1:
        if is_valid_number(i,i):
            dp[i] += dp[i-1]
        if is_valid_number(i-1,i):
            dp[i] += 1
    else:
        if is_valid_number(i,i):
            dp[i] += dp[i-1]
        if is_valid_number(i-1,i):
            dp[i] += dp[i-2]


print(dp[len(string)-1] % 1000000)